### PR TITLE
fix(history): promote distributed std to fp64 to avoid fp32 cancellation

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

`_compute_distributed_metrics` in `history.py` was computing
across-rank variance using the numerically unstable
`E[X²] - E[X]²` formula in **fp32**. For typical training metrics
(`tokens_per_sec ≈ 1e5`, `tflops ≈ 40`, `tokens_per_sec_per_gpu ≈ 1e4`)
the squared values are ~1e10 while the across-rank variance is
hundreds — that's 8+ orders of magnitude smaller than fp32 mantissa
precision allows, so the subtraction collapses to **exactly 0.0**.

## The symptom

Training logs where some metrics had `(±std)` and others were bare,
depending on whether the metric magnitude happened to fit inside
fp32's precision window:

```
[train] step=1 loss=2.467(± 0.17) ... tokens_per_sec=50982(± 111) tflops=15.75(± 0.035)
[train] step=2 loss=2.881(± 0.23) ... tokens_per_sec=122821(± 32) tflops=37.94          
[train] step=3 loss=2.584(± 0.14) ... tokens_per_sec=123881(± 45) tflops=38.27          
[train] step=4 loss=2.606(± 0.12) ... tokens_per_sec=124562(± 32) tflops=38.48(±5.5e-3)
```

Notice step 2-4 dropping `(±std)` on `tflops` despite per-rank
spread *guaranteed* to exist (every rank measured a slightly
different FLOPs throughput). Looked like ranks had identical values
when they were actually just losing the signal to fp32 precision.

## Numerical demo

```python
per_rank = [123880, 123920, 123870, 123890, 123900, 123860]
# OLD (fp32 E[X²] - E[X]²):     std=0.000000
# NEW (fp64 promotion):          std=19.720266
# torch.std(unbiased=False):     std=19.720266  ← matches exactly
```

## Fix

Promote `sum_vals` and `sq_vals` to fp64 before the all-reduce and
variance arithmetic. The max/min reductions stay in the original
dtype — they don't suffer cancellation.

```diff
- sum_vals = values.clone()
- sq_vals = values.square()
+ sum_vals = values.to(torch.float64)
+ sq_vals = sum_vals.square()
```

fp64 has ~16 sig digits of headroom which covers any realistic
training metric (even at 1e9 scale the variance arithmetic stays
stable). Cost is one extra cast + ~2x bandwidth on the squared-sum
all-reduce — negligible vs. the metric collection itself.

## Test

`tests/test_history.py::TestHistory::test_distributed_std_survives_fp32_cancellation`

Uses mocked `torch.distributed.all_reduce` to simulate a 6-rank
aggregation of the realistic `tokens_per_sec ≈ 1.24e5` values from
the failing log. Asserts `std > 0.0` with a regression message
pointing at the source file. Also verifies the std matches
`statistics.pstdev` to within 1e-3 relative tolerance — not just
"non-zero by accident".

**Verified the test catches the regression**: applied only the test
against the pre-fix source, observed the exact assertion failure:
> AssertionError: Distributed std collapsed to 0.0 — fp32 cancellation regressed

Applied the source fix, test passes.

## Test plan

- [x] `pytest tests/test_history.py` — 34/34 pass (was 33)
- [x] `pytest tests/test_history.py::TestHistory::test_distributed_std_survives_fp32_cancellation` against pre-fix source FAILS with expected message
- [x] Numerical demo: `std=19.72` matches `torch.std(unbiased=False)` exactly
- [ ] On a real multi-rank training run: `tflops(±X.XXX)` consistently has a non-zero std across steps

## Notes

- Bug is pre-existing — unrelated to PR #134 or #136. Caught while looking at training logs from the launch-flag work.
- The `clamp_min_(0.0)` on the variance line was the smoking gun — the original author knew about negative-variance edge cases and just clamped, treating it as "rounding error" when it was actually "lost the entire signal to fp32 cancellation".

## Summary by Sourcery

Ensure distributed metric standard deviation remains accurate for large-magnitude training metrics by improving numerical stability in the history aggregation logic.

Bug Fixes:
- Prevent distributed standard deviation for large-magnitude metrics from collapsing to zero due to fp32 catastrophic cancellation in variance computation.

Tests:
- Add a regression test that simulates multi-rank aggregation of large-magnitude metrics to verify distributed std is non-zero and matches the true population standard deviation.

### CODE REVIEW

File: src/ezpz/history.py [hunk 1]
```diff
@@ -1378,8 +1378,19 @@ def _compute_distributed_metrics(
...
+        # `E[X^2] - E[X]^2` doesn't lose its signal to fp32 cancellation.
+        # For typical training metrics (`tokens_per_sec ≈ 1e5`,
+        # `tflops ≈ 40`, etc.) the squared values are 8-10 orders of
+        # magnitude larger than the across-rank variance, and the
+        # subtraction collapses to exactly 0.0 in fp32 — `tflops/std=0.0`
+        # is then indistinguishable from "all ranks identical" and the
+        # console summary drops the `(±std)` parenthetical entirely.
+        # fp64 has ~16 sig digits of headroom which covers any realistic
+        # training metric. Cost: one extra cast + ~2x bandwidth on the
+        # squared-sum all-reduce, negligible vs. the metric collection.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/history.py]: Fixed in 3631edc. Updated comment to "two casts + ~2x bandwidth on **both** the sum and squared-sum all-reduces (max/min stay in the original dtype since they don't suffer cancellation)".

File: src/ezpz/history.py [hunk 2]
```diff
@@ -1378,8 +1378,19 @@ def _compute_distributed_metrics(
...
+        sum_vals = values.to(torch.float64)
+        sq_vals = sum_vals.square()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/history.py]: Fixed in 3631edc. Switched to `values.to(torch.float64, copy=True)` which forces a fresh tensor regardless of source dtype. Safe to follow with in-place all_reduce now even if `values` was already fp64.

File: tests/test_history.py [hunk 3]
```diff
@@ -189,6 +189,110 @@ def test_history_to_dataarray_handles_tuples_and_empty(self):
...
+                # sum of all rank contributions. The first time we
+                # see SUM the tensor is the rank-0 value; for the
+                # squared sum it's the rank-0 squared value. Re-derive
+                # from `arr` to keep this test agnostic of which call.
+                if torch.allclose(tensor.squeeze(), arr[0:1].to(tensor.dtype)):
+                    tensor.copy_(arr.sum().to(tensor.dtype).view_as(tensor))
+                else:
+                    tensor.copy_(
+                        arr.square().sum().to(tensor.dtype).view_as(tensor)
+                    )
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_history.py:265]: Fixed in 3631edc. Replaced the content-comparison dispatch with a `sum_call_count` list (used as a nonlocal-effective counter). Source always issues `SUM(sum_vals)` before `SUM(sq_vals)` so call-order dispatch is reliable, and we no longer break for inputs where `x ≈ x²`.

File: src/ezpz/history.py [hunk 4]
```diff
@@ -1378,8 +1378,19 @@ def _compute_distributed_metrics(
...
+        # For typical training metrics (`tokens_per_sec ≈ 1e5`,
+        # `tflops ≈ 40`, etc.) the squared values are 8-10 orders of
+        # magnitude larger than the across-rank variance, and the
+        # subtraction collapses to exactly 0.0 in fp32 — `tflops/std=0.0`
+        # is then indistinguishable from "all ranks identical" and the
+        # console summary drops the `(±std)` parenthetical entirely.
+        # fp64 has ~16 sig digits of headroom which covers any realistic
+        # training metric. Cost: one extra cast + ~2x bandwidth on the
+        # squared-sum all-reduce, negligible vs. the metric collection.
+        sum_vals = values.to(torch.float64)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/history.py]: Fixed in 3631edc. Wrapped the fp64 promotion + a trivial fp64 probe (`sum_vals.square().sum().item()`) in a try/except that falls back to the original dtype with a logged warning if fp64 isn't supported on the device.

The probe runs BEFORE any collectives — if I caught the exception mid-reduce, half the ranks would be in fp64 and half in fp32, which would deadlock the all_reduce. Probing locally first commits to a uniform dtype across the world.

Added `test_distributed_std_falls_back_when_fp64_unsupported` which monkeypatches `Tensor.to` to raise on fp64 (simulating an XPU device without fp64 support) and verifies the helper produces valid stats in fp32 rather than crashing the run.
